### PR TITLE
Add CODEOWNERS file

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

Per [discussion](https://docs.google.com/document/d/15JSFh3OMF9Iz6ov3q_xxGO_BL8hRnuse4IMUrqEIvcg/edit#bookmark=id.jheh6bf5f0vi) in the Astropy Dev telecon today, this PR adds a template CODEOWNERS file.

Maintainers can add themselves on an opt-in basis to get automatically requested for review on any PRs containing changes to their sub-package by creating a PR changing this file (or initially by requesting changes in this PR).

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by taldcroft** [comment]
@WilliamJamieson - nice. Please add me for `time`, `io.ascii` and `table`. Does this notify on draft PRs (hopefully not)?

**Comment by pllim** [comment]
No, it does not notify if the PR starts out as draft.

BUT if you start out with a real PR and then turn it to draft after the fact, then all bets are off.

**Comment by taldcroft** [comment]
>  BUT if you start out with a real PR and then turn it to draft after the fact, then all bets are off.

I can live with that.

File: CODEOWNERS.txt
Comment (astrofrog) [inline CODEOWNERS.txt:22]: Can we add an astropy/wcs/wcsapi entry here so that I can get assigned for reviews just for ``wcsapi`` but not all ``astropy/wcs``?

File: CODEOWNERS.txt
Comment (astrofrog) [inline CODEOWNERS.txt:21]: Happy to be added to visualization

File: CODEOWNERS.txt
Comment (astrofrog) [inline CODEOWNERS.txt:41]: Happy to be added to this one

File: CODEOWNERS.txt
Comment (nstarman) [inline CODEOWNERS.txt:6]:
```suggestion
# astropy/coordinates
astropy/cosmology           @nstarman
```

File: CODEOWNERS.txt
Comment (nstarman) [inline CODEOWNERS.txt:9]:
```suggestion
astropy/io/misc             @WilliamJamieson
astropy/io/registry         @nstarman
```

File: CODEOWNERS.txt
Comment (nstarman) [inline CODEOWNERS.txt:29]:
```suggestion
docs/cosmology              @nstarman
docs/io/ascii               @taldcroft
```

**Review by nstarman — COMMENTED** [review]
👍

File: CODEOWNERS.txt
Comment (Cadair) [inline CODEOWNERS.txt:44]: :raising_hand_man:

**Comment by pllim** [comment]
Thanks, all! When do we decide that this is good for a final review and merge?

**Review by nstarman — APPROVED** [review]

**Comment by matteobachetti** [comment]
Hi, I would like to be added to io.misc
Thanks!

File: CODEOWNERS.txt
Comment (hamogu) [inline CODEOWNERS.txt:2]:
```suggestion
# Despite the name of this file, the people listed here DO NOT OWN the code in astropy subpackages,
# the copyright is held by the project as a whole - see the license file for details. 
# 
# However, github currently (2022) does not offer a mechanism on their website to subscribe to
# a fraction of newly open PRs in astropy. Either one chooses the "watch" the entire repository and 
# receives  a large number of emails or one needs to manually check is new PRs for a certain sub-package
# exist. This file is a workaround for that problem:
# All names listed here by sub-packages will be requested for review for any new PR 
# and thus get an email notification.
# In many cases, the names below will be active sub-package maintainers, but anyone interested
# can open a PR to add their github handle to any subpackage they are interested in.


# Add your github username to opt-in to automatic review requests for
# specific sub-packages here:
```

**Comment by dhomeier** [comment]
Could you add me to io.ascii as well – thanks!

File: CODEOWNERS.txt
Comment (Cadair) [reply CODEOWNERS.txt:2]: I think you have a typo here:

> receives  a large number of emails or one needs to manually check is new PRs for a certain sub-package

I am not entirely sure what you want to say.

**Comment by astrofrog** [comment]
Out of curiosity why don't we just add all subpackage maintainers by default? We often assign them manually for review by default anyway right?

**Comment by pllim** [comment]
> Out of curiosity why don't we just add all subpackage maintainers by default?

The list is known to get outdated once in a while and not updated until after much debate. I know some maintainers have low capacity and do not respond to review requests and also there are some unfilled spots.

tl;dr -- staffing and politics.

**Comment by dhomeier** [comment]
I think the feeling in the devcon discussion was not to push anyone into getting more emails without their opting in.
But perhaps it's a good idea to ping the missing ones here again, as the Slack reminder may not have got everyone's attention.

File: CODEOWNERS.txt
Comment (dhomeier) [reply CODEOWNERS.txt:2]: "check if"? Maybe just shorten the whole thing a bit to
> Either one chooses to "watch" the entire repository and 
> \# thus receive a large number of emails or one needs to manually search for new PRs tagging a certain sub-package.

Comment (hamogu) [reply CODEOWNERS.txt:2]: I want to add an explanation of what this file is used for. It's not part of git, but just the gihub interface. Given the name of the file, anyone who find this file, might assume that the people listed there own the code. There are several aspects:
- if thinking of "own" as "legal possession": We have a license and that states the copyright, not this file.
- if thinking of "own" as "is responsible for": We curate astropy.org/team with the sub-package maintainers, but anyone is welcome to open PRs for any sub-package, even if they don't "own" it. That might be obvious to this crowed, but not necessarily to someone coming in new and browsing the repro.
- Since this this file is opt-in it will never by in sync (and likely will get out of sync more as we add and remove sub-package maintainers in the future). On the other hand, I see that people might want to sign up who are not necessarily the sub-package maintainers. I think that's OK and want to write that down so there is no misunderstanding.

**Comment by dhomeier** [comment]
> The list is known to get outdated once in a while and not updated until after much debate. I know some maintainers have low capacity and do not respond to review requests and also there are some unfilled spots.

I am also a bit confused at this point – are you referring to the subpackage maintainers list here? Being an official list of roles it does of course take a somewhat longer process to update, but it also is in a way the authoritative list of people responsible for the subpackages, so I would also always refer to that, even if not always sending requests to every co-maintainer. But for anyone not familiar with the infrastructure that's often the only point to turn to for a review, if they do not want to rely on github's suggestions of people who have edited related files somewhat recently.

I understand this file as an additional resource to take some load off maintainers with limited review capacity, but it not as a substitute for keeping the maintainer's list as up-to-date as possible, especially with a number of packages still not covered here.

A question that just came to my mind: will the "codeowners" from this file be visible to the submitter as "requested for review" on their PR? In that case we might recommend the following procedure:
1. Request their own reviewer if they already have someone in mind.
2. Check for reviewer automatically added via this file.
3. If there still are too few reviewers, or no reaction after some time, add the sub-package maintainers (it would still be helpful to have an easier way to locate them other than going to the Project-Team webpage).

File: CODEOWNERS.txt
Comment (dhomeier) [reply CODEOWNERS.txt:15]: Typos still need to be addressed.

**Comment by bsipocz** [comment]
> I also worry that this list is going to get out of date with active roles so it would be better if it was auto generated from roles ideally? I would have thought a primary duty of a subpackage maintainer should be to review PRs to their subpackage?

just another reminder that all roles should be managed as github teams rather than as individuals, and it should be a top priority for "management" to know about and use the technical solutions that are already out there with the purpose of making things easier for maintainers. 
(and there is a good example of e.g. sunpy of how to use this file).

**Comment by hamogu** [comment]
> just another reminder that all roles should be managed as github teams rather than as individuals

Astropy uses github teams for permissions such as who can write to a repro and I think they work well for that;  I think all of the people currently looking at this file are aware of github teams. In this case, however, I personally prefer to have individual handles. We want to functionality for a user to opt-in. We *could* have that fine-grained controls with more teams (@io-ascii-review-opt-in, @io-fits-review-opt-in, etc.). Anyone can open a PR to add or remove themselves if they wish, making it easier for people to chose to be e.g. alerted of changes in just some of several packages they maintain or even for packages they don't maintain, but are just interested in. 

If, after some trial period, it turns out that this essentially becomes a list of sub-package maintainers, then I suggest to revisit this and possibly change it to what @bsipocz suggests, but I would prefer to see how it goes first. At the dev-telecon where we discussed this, there was strong sentiment that we should not auto-subscribe all maintainers for their sub-packages (which seems to be how sunpy does it).

File: CODEOWNERS.txt
Comment (nstarman) [inline CODEOWNERS.txt:20]:
```suggestion
astropy/cosmology           @astropy/cosmology
```

File: CODEOWNERS.txt
Comment (nstarman) [inline CODEOWNERS.txt:43]:
```suggestion
docs/cosmology              @astropy/cosmology
```

**Review by nstarman — COMMENTED** [review]
The cosmology team has been moved to be a sub-team of astropy-core-maintainers and can now review PRs. Hopefully my last comment. Thanks @WilliamJamieson.

File: CODEOWNERS.txt
Comment (WilliamJamieson) [reply CODEOWNERS.txt:20]: Maybe other groups of sub-package maintainers should also form teams like this and as a unit decide if they want notifications like this?

**Comment by WilliamJamieson** [comment]
To bring my https://github.com/astropy/astropy/pull/13241#discussion_r875236630, into the general discussion here. Are there any other groups of sub-package maintainers who wish to form one of these sub-teams and collectively register themselves in this?

File: CODEOWNERS.txt
Comment (nstarman) [reply CODEOWNERS.txt]: I personally think it's a good idea, but pragmatically it hardly makes a difference since most teams are only a few people. Teams have more longevity than individual maintainers, and there are nice options like review load balancing, but again, not much practical difference.

**Comment by nstarman** [comment]
@WilliamJamieson tests are hanging. Maybe rebase?

**Comment by WilliamJamieson** [comment]
> @WilliamJamieson tests are hanging. Maybe rebase?

Done

**Comment by pllim** [comment]
From the dev telecon today, here are the 3 things that need follow-up. Once they are done, I can override the "ci skip" stuff and merge it since this PR does not need CI.

1. Fix typos, as pointed out above.
2. Remove ".txt" (unnecessary).
3. Squash the commits.

**Comment by pllim** [comment]
After merge, we can actually see how well this works for the people who opt-in and go from there.

**Pull request merged**
